### DISCUSSION TITLE
Losses and loss functions

### DISCUSSION THREAD

**Discussion by hy23-lfb** [voxelmorph/voxelmorph#466] [Q&A]
Hi @adalca,
First of all, I would like to thank you and your team for your contribution and continuous support. Discussion and Issue forums have been very helpful.

I would like to ask you about the three losses I see at the end of each epoch, i.e. loss, vxm_dense_transformer_loss and vxm_dense_flow_resize_loss.

1. What are these 3 different losses? And what are the references considered? For e.g., I believe "loss" is the main loss computed between registered image and the fixed image. So, similarly what are the references considered for computing vxm_dense_transformer_loss  and vxm_dense_flow_resize_loss?
2. Is it also the case that the loss function used to calculate all these losses is the same? For example, if I say in my train.py script that my loss function is "mse", is it the case that all these losses are calculated with "mse"? I am interested in the scenario where I want the main loss to be calculated with "mutual information", but the transformer loss to be calculated with "mse". Is this possible? If so, any advice on how to achieve this would be extremely helpful.

Many thanks in advance,
Harsha

**Comment by adalca** [comment]
1. This depends on a few things, like which model you're running and where you see these losses, but I believe what you're referring to is actually just two losses (img_matching, and smoothness), and the `loss` is just the weighted sum.

2. Sorry I'm not quite sure I follow what you mean by computing the 'main' loss with MI and the transformer loss with 'mse'. The 'mse' in train refers to the image matching term only.

**Reply by hy23-lfb** [reply]
First of all, thank you very much for taking the time to answer my question. And please also excuse my unclear question above, I will try to be more specific and clear now.

When I train the model, I see the following losses in each epoch.
```
11/560 [..............................] - ETA: 9:28 - loss: 0.0044 - vxm_dense_transformer_loss: 0.0035 - vxm_dense_flow_loss: 2.1589e-11
```
When I spoke of 3 losses, I meant `loss`, `vxm_dense_transformer_loss` and `vxm_dense_flow_loss`.

And when I meant 'main' loss, I was referring to `loss`, which is the weighted sum of `img_matching` and `smoothness`. Thanks also for the confirmation. But I did not understand what the other two losses are, namely `vxm_dense_transformer_loss` and `vxm_dense_flow_loss`. 

So my questions are about these two losses: -.

1. In order for losses to be calculated, there must be a ground truth (_reference values_ in my question above) - what is the ground truth for these two losses?
2. In order for losses to be calculated, we also need to specify a loss function for losses - what loss function is used in calculating the loss values for these two losses?
3. Is it possible that I compute `loss` with mutual information, but `vxm_dense_transformer_loss` with mse?

**Comment by adalca** [comment]
1. voxelmorph is an _unsupervised_ method, so no ground truth needed. Please see the voxelmorph papers (https://arxiv.org/abs/1809.05231) for details, but basically the first loss is for example MSE(x_1, warped(x_2)), where x_1 and x_2 are the inputs.   
2. by default it's MSE and vxm.losses.Grad
3. you could use vxm.losses.MI instead of MSE for the first loss.
